### PR TITLE
mkldnn: relu, sum and batch norm

### PR DESCRIPTION
**Pull request by sreekanth-yalachigere**
**Activator Relu (eltwise)**

```
core/providers/mkldnn/activation
    activations.cc
    activations.h
```
Modified CPU Relu class definition to enable MKLDNN Relu class to inherit for fallback option

`class Relu final : public OpKernel { 
`to
`class Relu : public OpKernel {`

**Sum Kernel**

```
core\providers\mkldnn\math\sum.cc
core\providers\mkldnn\math\sum.h
```
Modified base class `Sum_6 `in `core\providers\cpu\math\element_wise_ops.h` Removed final
for mkldnn to fallback to cpu computation.

**Batch Norm Kernel**

```
core\providers\mkldnn\nn\batch_norm.cc
core\providers\mkldnn\nn\batch_norm.h
```
Modified base class `onnxruntime::BatchNorm` in file `core\providers\cpu\nn\batch_norm.h`
Removed final and made epsilon_ as protected member.

### CODE REVIEW

**Comment by sreekanth-yalachigere** [comment]
@snnn @@jywu-msft 

Relu uninitialized memory warning fixed: MKLDNN V0.17.2
Sum: unexpected result: Fixed

File: onnxruntime/core/providers/mkldnn/math/sum.cc [hunk 1]
```diff
@@ -0,0 +1,231 @@
...
+namespace {
+// Struct which encapsulates parameters for MKLDNN Sum primitives.
+struct SumParams {
+  const std::vector<mkldnn::memory::dims>& src_dims;
+  const mkldnn::memory::dims& dst_dim;
+  const int num_inputs;
+  const int num_dimensions;
+
+  SumParams(const std::vector<mkldnn::memory::dims>& dims,
+    const mkldnn::memory::dims& dst_dims, const int numinputs,
```
Comment (jywu-msft) [inline onnxruntime/core/providers/mkldnn/math/sum.cc]: can remove const from numinputs and dimensions parameters.

File: onnxruntime/core/providers/mkldnn/activation/activations.cc [hunk 2]
```diff
@@ -0,0 +1,204 @@
...
+namespace mkl_dnn {
+
+namespace {
+// Struct which encapsulates parameters for MKLDNN Pool primitive.
+struct ReluParams {
+  mkldnn::memory::dims& src_dims;
+  mkldnn::memory::dims& dst_dims;
+  size_t num_dimensions;
+
+  ReluParams(mkldnn::memory::dims& src_dims, mkldnn::memory::dims& dst_dims,
```
Comment (jywu-msft) [inline onnxruntime/core/providers/mkldnn/activation/activations.cc]: src_dims and dst_dims references can be made const

File: onnxruntime/core/providers/mkldnn/activation/activations.cc [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+#include "core/providers/mkldnn/activation/activations.h"
+#include "core/providers/mkldnn/mkldnn_fwd.h"
+
+namespace onnxruntime {
+namespace mkl_dnn {
+
+namespace {
+// Struct which encapsulates parameters for MKLDNN Pool primitive.
+struct ReluParams {
+  mkldnn::memory::dims& src_dims;
```
Comment (jywu-msft) [inline onnxruntime/core/providers/mkldnn/activation/activations.cc]: src_dims and dst_dims references can be made const

File: onnxruntime/core/providers/mkldnn/nn/batch_norm.cc [hunk 4]
```diff
@@ -0,0 +1,287 @@
...
+  const mkldnn::memory::dims& var_dims;
+  const mkldnn::memory::dims& dst_dims;
+  const float epsilon;
+  const int num_dimensions;
+
+  BatchNormParams(const mkldnn::memory::dims& src_dims_mkl,
+    const mkldnn::memory::dims& scale_dims_mkl,
+    const mkldnn::memory::dims& b_dims_mkl, const mkldnn::memory::dims& mean_dims_mkl,
+    const mkldnn::memory::dims& var_dims_mkl, const mkldnn::memory::dims& dst_dims_mkl,
+    const float eps, const int dimensions)
```
Comment (jywu-msft) [inline onnxruntime/core/providers/mkldnn/nn/batch_norm.cc]: eps and dimensions params don't need to be const as they are not references and will be copied.

**Comment by sreekanth-yalachigere** [comment]
Re review: done.

File: onnxruntime/core/providers/mkldnn/activation/activations.cc [hunk 5]
```diff
@@ -0,0 +1,204 @@
...
+
+namespace {
+// Struct which encapsulates parameters for MKLDNN Pool primitive.
+struct ReluParams {
+  const mkldnn::memory::dims& src_dims;
+  const mkldnn::memory::dims& dst_dims;
+  size_t num_dimensions;
+
+  ReluParams(const mkldnn::memory::dims& src_dims, const mkldnn::memory::dims& dst_dims, 
+		size_t dimensions = 0)
```
Comment (pranavsharma) [inline onnxruntime/core/providers/mkldnn/activation/activations.cc]: is dimensions intentionally a default parameter?

File: onnxruntime/core/providers/mkldnn/nn/batch_norm.h [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+#include "core/framework/op_kernel.h"
+#include "core/providers/cpu/nn/batch_norm.h"
+
+namespace onnxruntime {
+namespace mkl_dnn {
+
+template <typename T>
+class BatchNorm final : public onnxruntime::BatchNorm<T> {
+ public:
+  BatchNorm(const OpKernelInfo& info) : onnxruntime::BatchNorm<T>(info) {}
```
Comment (pranavsharma) [inline onnxruntime/core/providers/mkldnn/nn/batch_norm.h]: Mark single parameter constructors explicit.

File: onnxruntime/core/providers/mkldnn/math/sum.h [hunk 7]
```diff
@@ -0,0 +1,22 @@
...
+#include "core/providers/cpu/math/element_wise_ops.h"
+#include "../mkldnn_execution_provider.h"
+
+namespace onnxruntime {
+namespace mkl_dnn {
+
+template <typename T>
+class Sum final : public onnxruntime::Sum_6<T> {
+ public:
+   Sum(const OpKernelInfo& info) : onnxruntime::Sum_6<T>(info) {}
```
Comment (pranavsharma) [inline onnxruntime/core/providers/mkldnn/math/sum.h]: Mark single parameter constructors explicit.

File: onnxruntime/core/providers/mkldnn/activation/activations.h [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+#include "core/framework/op_kernel.h"
+#include "core/providers/cpu/activation/activations.h"
+
+namespace onnxruntime {
+namespace mkl_dnn {
+
+template <typename T>
+class Relu : public onnxruntime::Relu<T> {
+ public:
+  Relu(const OpKernelInfo& info) : onnxruntime::Relu<T>(info) {}
```
Comment (pranavsharma) [inline onnxruntime/core/providers/mkldnn/activation/activations.h]: Mark single parameter constructors explicit.

File: onnxruntime/core/providers/mkldnn/activation/activations.cc [hunk 9]
```diff
@@ -0,0 +1,204 @@
...
+
+template <typename T>
+Status Relu<T>::Compute(OpKernelContext* context) const {
+  const Tensor* X = context->Input<Tensor>(0);
+  Tensor* Y = context->Output(0, X->Shape());
+  
+  const TensorShape& x_shape = X->Shape();
+  const auto& x_dims = x_shape.GetDims();
+
+  if (X->Shape().NumDimensions() > 5 ) {
```
Comment (pranavsharma) [inline onnxruntime/core/providers/mkldnn/activation/activations.cc]: can you leave a comment here saying why we're defaulting to the non-mkldnn implementation?

File: onnxruntime/core/providers/mkldnn/activation/activations.cc (hunk 5 above)
Comment (jywu-msft) [reply onnxruntime/core/providers/mkldnn/activation/activations.cc]: actually for dimensions, can't we compute it from the src_dims? same comment for all these kernels...

File: onnxruntime/core/providers/mkldnn/activation/activations.cc [hunk 10]
```diff
@@ -0,0 +1,204 @@
...
+
+  const T* src_data = X->template Data<T>();
+  T* dst_data = Y->template MutableData<T>();
+
+  mkldnn::memory::dims src_dims_mkl(x_dims.begin(), x_dims.end());
+  mkldnn::memory::dims dst_dims_mkl(y_dims.begin(), y_dims.end());
+
+  try {
+    ReluParams pool_params(src_dims_mkl, dst_dims_mkl, x_shape.NumDimensions());
+    ReluPrimitive<T>* relulPrimitive = ReluPrimitivePool<T>::Get(pool_params);
```
Comment (pranavsharma) [inline onnxruntime/core/providers/mkldnn/activation/activations.cc]: fix naming convention and typo.

File: onnxruntime/core/providers/mkldnn/activation/activations.cc [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+		  static_cast<void*>(const_cast<T*>(dst_data)));
+	  context_.stream->submit(context_.net);
+
+	  context_.src_mem->set_data_handle(nullptr);
+	  context_.dst_mem->set_data_handle(nullptr);
+	  return;
+  }
+
+  std::unique_ptr<mkldnn::memory::desc> 
+	GetDstMemoryDesc() const { return context_.dst_md; }
```
Comment (pranavsharma) [inline onnxruntime/core/providers/mkldnn/activation/activations.cc]: this is not used any where. get rid of dead functions.

File: onnxruntime/core/providers/mkldnn/activation/activations.cc [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+
+  std::unique_ptr<mkldnn::eltwise_forward::primitive_desc> 
+	GetPrimitiveDesc() const {
```
Comment (pranavsharma) [inline onnxruntime/core/providers/mkldnn/activation/activations.cc]: is this used anywhere? if not, remove it.

File: onnxruntime/core/providers/mkldnn/math/sum.cc [hunk 13]
```diff
@@ -0,0 +1,231 @@
...
+};
+} // namespace_
+
+template <typename T>
+Status Sum<T>::Compute(OpKernelContext* context) const {
+  int num_inputs = static_cast<int>(OpKernel::Node().InputDefs().size());
+
+  ORT_ENFORCE(num_inputs > 0, "MKLDNN Sum kernel: Must have at least one input");
+
+  if (num_inputs == 1) {
```
Comment (pranavsharma) [inline onnxruntime/core/providers/mkldnn/math/sum.cc]: please make a note why you're defaulting to non-mkldnn implementation here.

File: onnxruntime/core/providers/mkldnn/nn/batch_norm.cc [hunk 14]
```diff
@@ -0,0 +1,287 @@
...
+namespace {
+// Struct which encapsulates parameters for MKLDNN BatchNorm primitive.
+struct BatchNormParams {
+  const mkldnn::memory::dims& src_dims;
+  const mkldnn::memory::dims& scale_dims;
+  const mkldnn::memory::dims& b_dims;
+  const mkldnn::memory::dims& mean_dims;
+  const mkldnn::memory::dims& var_dims;
+  const mkldnn::memory::dims& dst_dims;
+  const float epsilon;
```
Comment (pranavsharma) [inline onnxruntime/core/providers/mkldnn/nn/batch_norm.cc]: why are epsilon and num_dimensions passed as const?

File: onnxruntime/core/providers/mkldnn/nn/batch_norm.cc [hunk 15]
```diff
@@ -0,0 +1,287 @@
...
+      : cpu_engine_(GetEngine()) {
+    context_.stream.reset(new mkldnn::stream(mkldnn::stream::kind::eager));
+    if (context_.batchnorm_fwd == nullptr) {
+      Initialize(params);
+    }
+  }
+
+  ~BatchNormPrimitive() = default;
+
+  void Compute(const T* src_data, const T* scale_data, const T* b_data,
```
Comment (pranavsharma) [inline onnxruntime/core/providers/mkldnn/nn/batch_norm.cc]: please follow similar guidelines as Relu. data that is intended to be mutated (like output data) should not be passed as const to begin with (to avoid const_cast later on).

File: onnxruntime/core/providers/mkldnn/nn/batch_norm.cc [hunk 16]
```diff
@@ -0,0 +1,287 @@
...
+    context_.src_mem->set_data_handle(
+      static_cast<void*>(const_cast<T*>(src_data)));
+    context_.mean_mem->set_data_handle(
+      static_cast<void*>(const_cast<T*>(mean_data)));
+    context_.var_mem->set_data_handle(
+      static_cast<void*>(const_cast<T*>(var_data)));
+    context_.dst_mem->set_data_handle(
+      static_cast<void*>(const_cast<T*>(dst_data)));
+
+    T* scaleShift_buf = static_cast<T*>(context_.scale_shift_mem->get_data_handle());
```
Comment (pranavsharma) [inline onnxruntime/core/providers/mkldnn/nn/batch_norm.cc]: fix scaleShift naming convention.

**Comment by sreekanth-yalachigere** [comment]
@pranavsharma @jywu-msft 
thanks for your review. addressing your review comments.

File: onnxruntime/core/providers/mkldnn/activation/activations.cc [hunk 17]
```diff
@@ -0,0 +1,204 @@
...
+      : cpu_engine_(GetEngine()) {
+    context_.stream.reset(new mkldnn::stream(mkldnn::stream::kind::eager));
+    if (context_.relu_fwd == nullptr) {
+      Initialize(params);
+    }
+  }
+
+  ~ReluPrimitive() = default;
+
+  void Compute(const T* src_data, const T* dst_data) {
```
[the remark this answers is not in the document]
Comment (sreekanth-yalachigere) [reply onnxruntime/core/providers/mkldnn/activation/activations.cc]: @pranavsharma I looked at compute class of other kernels (conv, pool and lrn). In all cases, dst_data is const and const_cast is used.

[the remark this answers is not in the document]
Comment (pranavsharma) [reply onnxruntime/core/providers/mkldnn/activation/activations.cc]: it's rather unfortunate that this was missed in our code reviews before. we should do the right thing now.

[the remark this answers is not in the document]
Comment (sreekanth-yalachigere) [reply onnxruntime/core/providers/mkldnn/activation/activations.cc]: Can I leave it as const now? 
I'll make these changes in my next PR.

[the remark this answers is not in the document]
Comment (pranavsharma) [reply onnxruntime/core/providers/mkldnn/activation/activations.cc]: We should change the incorrect usage in this code. I'll fix the previous violations.

**Comment by sreekanth-yalachigere** [comment]
addressed all the issues. 
removed num_dimensions and using src_dims.size() instead.
removed const _cast and const from dst_data
explicit for single parameter constructors
typo and naming conversions

File: onnxruntime/core/providers/mkldnn/activation/activations.cc [hunk 18]
```diff
@@ -178,12 +169,12 @@ Status Relu<T>::Compute(OpKernelContext* context) const {
...
-
+  
   try {
-    ReluParams pool_params(src_dims_mkl, dst_dims_mkl, x_shape.NumDimensions());
-    ReluPrimitive<T>* relulPrimitive = ReluPrimitivePool<T>::Get(pool_params);
+    ReluParams pool_params(src_dims_mkl, dst_dims_mkl);
+    ReluPrimitive<T>* relul_primitive = ReluPrimitivePool<T>::Get(pool_params);
 
-    relulPrimitive->Compute(src_data, dst_data);
+    relul_primitive->Compute(src_data, dst_data);
```
Comment (pranavsharma) [inline onnxruntime/core/providers/mkldnn/activation/activations.cc]: typo: relu_primitive.

Comment (sreekanth-yalachigere) [reply onnxruntime/core/providers/mkldnn/activation/activations.cc]: sorry. Missed it.

**Comment by pranavsharma** [comment]
LGTM 👍

**Comment by pranavsharma** [comment]
@jywu-msft i guess this can be merged, unless you've any comments.
